src/dir-hamt-sharded.js

```diff
@@ -27,41 +32,180 @@ function shardedDirExporter (cid, node, name, path, pathRest, resolve, size, dag
+        map((link) => {
+          // remove the link prefix (2 chars for the bucket index)
+          const entryName = link.name.substring(2)
+          const entryPath = entryName ? path + '/' + entryName : path
 
-        if (accept) {
           return {
-            depth: p ? depth + 1 : depth,
-            name: p,
-            path: pp,
+            depth: entryName ? depth + 1 : depth,
+            name: entryName,
+            path: entryPath,
             multihash: link.cid.buffer,
-            pathRest: p ? pathRest.slice(1) : pathRest,
+            pathRest: entryName ? pathRest.slice(1) : pathRest,
             parent: dir || parent
           }
-        } else {
-          return ''
-        }
-      }),
-      filter(Boolean),
-      resolve
-    )
-  ]
-
-  // place dir before if not specifying subtree
-  if (!pathRest.length || options.fullPath) {
+        }),
+        resolve
+      )
+    ]
+
+    // place dir before if not specifying subtree
     streams.unshift(values([dir]))
+
+    return cat(streams)
+  }
+
+  const deferred = defer.source()
+  const targetFile = pathRest[0]
+
+  // recreate our level of the HAMT so we can load only the subshard in pathRest
+  waterfall([
+    (cb) => {
+      if (!options.rootBucket) {
+        options.rootBucket = new Bucket({
+          hashFn: DirSharded.hashFn
+        })
+        options.hamtDepth = 1
+
+        return addLinksToHamtBucket(node.links, options.rootBucket, options.rootBucket, cb)
+      }
+
+      return addLinksToHamtBucket(node.links, options.lastBucket, options.rootBucket, cb)
+    },
+    (cb) => findPosition(targetFile, options.rootBucket, cb),
+    (position, cb) => {
+      let prefix = toPrefix(position.pos)
+      const bucketPath = toBucketPath(position)
+
+      if (bucketPath.length > (options.hamtDepth)) {
+        options.lastBucket = bucketPath[options.hamtDepth]
+
+        prefix = toPrefix(options.lastBucket._posAtParent)
+      }
+
+      const streams = [
+        pull(
+          values(node.links),
+          map((link) => {
+            const entryPrefix = link.name.substring(0, 2)
+            const entryName = link.name.substring(2)
+            const entryPath = entryName ? path + '/' + entryName : path
+
+            if (entryPrefix !== prefix) {
+              // not the entry or subshard we're looking for
+              return false
+            }
+
+            if (entryName && entryName !== targetFile) {
+              // not the entry we're looking for
+              return false
+            }
+
+            if (!entryName) {
+              // we are doing to descend into a subshard
+              options.hamtDepth++
+            } else {
+              // we've found the node we are looking for, remove the context
+              // so we don't affect further hamt traversals
+              delete options.rootBucket
+              delete options.lastBucket
+              delete options.hamtDepth
+            }
```

Comment:
Wait so...is the change of the values of these `options` depended on outside of `shardedDirExporter`? That doesn't sound good...

Reply:
No, but each invocation of the `shardedDirExporter` is invoked independently of the others.  The hashing algorithm used by HAMT needs to know about more than the current level because the buckets aren't related to directories so we need some way of passing that information along during a graph traversal.

Once we've found the target node we can remove the data as we're done with the traversal.  We want to remove it in case we traverse from a HAMT to a normal dir and then back to a HAMT.

Since a HAMT is only made up of sub shards, files and directories, once we hit a file or directory we are done.  That is, there's no chance of us going from one HAMT directly to another HAMT in a traversal, even if the directory we resolve to happens to be a HAMT (since it's a directory and not a sub shard).

Reply:
Ok understood - do you think there's an improvement to be made here in terms of understandability? Can we raise an issue for a future resolution?

Reply:
It's kind of bad form to modify the incoming arguments anyway.  When we do the refactor in #7 it'd be a good time to revisit this as I don't think the original approach expected any traversals to require state.